Implemented viewset extra action for rating entries

Stanislav-Lykhachov opened the pull request:
Business logic function was created in services.py
Value validation for marks was has been implemented.
Serializer takes only mark value.

fix #3 
fix #1

Commit:
Implement viewset extra action for rating entries

Business logic function was created in services.py
Value validation for marks was has been implemented.
Serializer takes only mark value.

vladyslav-bezpalko commented on littleblog/entry/models.py:
```diff
@@ -41,10 +41,10 @@ class Entry(models.Model):
     topic = models.ManyToManyField(Topic, related_name='entries')
     sum_of_marks = models.IntegerField(default=0)
-    current_rate = models.FloatField(default=0)
+    current_rating = models.FloatField(default=0)
```
You are changing DB schema here - DB migration file has to be included in PR

Stanislav-Lykhachov replied:
Hm. migration folder was tracking by git (so I supposed all the files created there would be added), but this file was really untracked.

vladyslav-bezpalko commented on littleblog/entry/serializers.py:
```diff
@@ -58,13 +58,25 @@ class Meta:
+    current_rating = serializers.FloatField()
+    amount_of_marks = serializers.IntegerField()
+
+    def validate_mark(self, value):
+        if value < 1 or value > 5:
+            raise serializers.ValidationError('Mark value must be between 1 and 5.')
+        return value
```
1. drf has something for your case. 

```
mark = serializers.IntegerField(min_value=..., max_value=..., ...)
```

2. https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad

Stanislav-Lykhachov replied:
But how can I set my own error text now?  Probably it doesn't matter and it's not needed in this simple case...

vladyslav-bezpalko replied:
Yes, in our case custom error message is not a requirement. But, for the future:
![Screenshot 2022-05-09 at 15 20 59](https://user-images.githubusercontent.com/74052345/167408660-948083fd-f546-41b7-96a8-2e2edc7c78aa.png)

You can pass `error_messages` parameter in field definition. Like this `error_messages={"min_value": "Mark value must be bigger than {min_value}"}`. 
However, your current coded uses two values in error messages. To implement it like that you would need to override IntegerField with custom code and handle error message in any way you want

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -1,13 +1,16 @@
 
+from .serializers import EntryCreateSerializer, UserProfileCreateSerializer, RatingUpdateSerializer, EntryDetailSerializer
+from .models import Entry
+from rest_framework.permissions import IsAuthenticated
```
wrong import groupping

Stanislav-Lykhachov replied:
I forgot that IsAuthenticated is not custom permission. Already moved to the top group

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -20,37 +23,17 @@ class EntryViewSet(viewsets.ModelViewSet):
 
+    @action(detail=True, methods=['patch'])
+    def rate(self, request, pk=None):
+        instance = self.get_object()
```
I'd suggest naming this var `entry` - it'd be much easier to read this code. `instance` is too broad, can be anything.

Stanislav-Lykhachov replied:
Fixed.

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -20,37 +23,17 @@ class EntryViewSet(viewsets.ModelViewSet):
+        instance = self.get_object()
+        serializer = RatingUpdateSerializer(instance, data=request.data, partial=True)
+        if serializer.is_valid(raise_exception=True):
+            set_mark(serializer.instance, **serializer.validated_data)  # rating logic storing in services.py
+            serializer.save()
```
call `instance.save()` in service function, ignore `serializer.save()`

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -20,37 +23,17 @@ class EntryViewSet(viewsets.ModelViewSet):
+    def rate(self, request, pk=None):
+        instance = self.get_object()
+        serializer = RatingUpdateSerializer(instance, data=request.data, partial=True)
+        if serializer.is_valid(raise_exception=True):
```
no need to do `if` when you use `raise_exception=True`.

It makes little sense, because when an exception is raised - the code won't go further anyway. ie, this code works same, but more straightforward:
```
serializer.is_valid(raise_exception=True)
set_mark(instance, **serializer.validated_data)
```

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -20,37 +23,17 @@ class EntryViewSet(viewsets.ModelViewSet):
         if self.action == 'create':
             return EntryCreateSerializer
         return EntryDetailSerializer
-
-
-class RateUpdateView(generics.UpdateAPIView):
-    """
-    Вот это есть та самая вьюшка, которая отвечает за подсчет рейтинга записи.
-    В url нужно дописать в виде параметров pk и mark (оценка, которую ставишь) и отправить PATCH request
-    """
-
-    permission_classes = (IsAuthenticated,)
-    serializer_class = RateUpdateSerializer
-
-    def update(self, request, *args, **kwargs):
-        instance = self.get_object()
-        instance.sum_of_marks += int(request.query_params['mark'])
-        instance.amount_of_marks += 1
-        instance.current_rate = instance.sum_of_marks / instance.amount_of_marks
-        instance.save()
-
-        if getattr(instance, '_prefetched_objects_cache', None):
-            instance._prefetched_objects_cache = {}
-
-        return Response(self.serializer_class(instance).data)
-
-    def get_queryset(self):
-        return Entry.objects.all()
-
-    def get_object(self):
-        queryset = self.filter_queryset(self.get_queryset())
-        obj = generics.get_object_or_404(queryset, pk=self.request.query_params['pk'])
-        return obj
```
🔥 👍🏼 

The new implementation looks much more elegant, yeah?

Stanislav-Lykhachov replied:
Yeah, I like it's new look : D

vladyslav-bezpalko commented on littleblog/entry/views.py:
```diff
@@ -20,37 +23,17 @@ class EntryViewSet(viewsets.ModelViewSet):
     queryset = Entry.objects.all()
     permission_classes = [IsAuthenticated, IsOwnerOrReadOnly]
 
+    @action(detail=True, methods=['patch'])
```
For "additonal API actions" better use `POST`. `PATCH` is almost always about the edit. Yes, we do edit under the hood, but for API users this action is "do smth" (POST) rather than "edit smth" (PUT/PATCH)

Commit:
Add migration file

Commit:
Change hardcoded mark field validation to
more appropriate look

Commit:
Improve rating action

Action method was changed to POST
Saving of modified Entry instance was simplified

2DLenin commented:
@vladyslav-bezpalko All the fixes are done

Commit:
change import list

vladyslav-bezpalko commented on littleblog/entry/services.py:
```diff
@@ -0,0 +1,5 @@
+    instance.sum_of_marks += kwargs['mark']
+    instance.amount_of_marks += 1
+    instance.current_rating = instance.sum_of_marks / instance.amount_of_marks
+    instance.save()
```
pep8: missing newline on the end of the file

Stanislav-Lykhachov replied:
@vladyslav-bezpalko New line has been added.

Stanislav-Lykhachov committed:
Add empty line
